File: src/BirthdayService.ts

import { ProductionCustomersRepository } from './customers/ProductionCustomersRepository.js'
import { DiscountCodeGenerator } from './discount/DiscountCodeGenerator.js'
import { ProductionEmailSender } from './email/ProductionEmailSender.js'
import { ProductionLogger } from './logger/ProductionLogger.js'

export class BirthdayService {
  private readonly customerRepository: ProductionCustomersRepository
  private readonly emailSender: ProductionEmailSender
  private readonly logger: ProductionLogger

  constructor(
    customerRepository: ProductionCustomersRepository,
    emailSender: ProductionEmailSender,
    logger: ProductionLogger,
  ) {
    this.customerRepository = customerRepository
    this.emailSender = emailSender
    this.logger = logger
  }

  greetCustomersWithBirthday(today: Date) {
    const customers = this.customerRepository.findWithBirthday(today)
    customers.forEach((customer) => {
      const discountCode = new DiscountCodeGenerator().generate()
      const template =
        'Happy birthday, {name}! Here is your discount code: {discount}'.replace(
          '{discount}',
          discountCode.getCode(),
        )
      customer.sendEmail(template, this.emailSender)
      this.logger.log('INFO', customer.fillWithEmail('Email sent to {email}'))
    })
  }
}

File: src/main/java/BirthdayService.java
package main.java;

import java.util.Date;

public class BirthdayService {
    private final ProductionCustomersRepository customerRepository;
    private final EmailSender emailSender;
    private final ProductionLogger logger;

    public BirthdayService(
            ProductionCustomersRepository customerRepository,
            EmailSender emailSender,
            ProductionLogger logger) {
        this.customerRepository = customerRepository;
        this.emailSender = emailSender;
        this.logger = logger;
    }

    public void greetCustomersWithBirthday(Date today) {
        var customers = customerRepository.findWithBirthday(today);
        for (Customer customer : customers) {
            var discountCode = new DiscountCodeGenerator().generate();
            var template = "Happy birthday, {name}! Here is your discount code: {discount}"
                    .replace("{discount}", discountCode.getCode());
            customer.sendEmail(template, emailSender);
            logger.log("INFO", customer.fillWithEmail("Email sent to {email}"));
        }
    }
} 